testers.testBuildFailure: fix structuredAttrs support

MattSturgeon opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

I noticed that `testers.testBuildFailure` was not working with `__structuredAttrs`.

The main issues are, when `structuredAttrs` is enabled:
- `$__structuredAttrs` is not set
- `$outputs` is not set
- `$outputs` defined by `NIX_ATTRS_SH_FILE` is an associative array
  - This means it is unordered, so we cannot determine the "default" output

As per:
https://github.com/NixOS/nixpkgs/blob/ecd810c4cc90890229ef353a12fb384c47ee75c3/pkgs/stdenv/generic/default-builder.sh#L1-L3

- `$outputs` not being set is fixed by sourcing `$NIX_ATTRS_SH_FILE`
- `$__structuredAttrs` not being set is fixed by sourcing `setup.sh`
- `$outputs` being unordered is worked-around by passing in `head orig.outputs` using `replaceVars`

#### Test using:
```sh
nix-build -A tests.testers.testBuildFailure
```

<details><summary>Original Description</summary>
<p>


The main issue was that when structuredAttrs is enabled, none of the attr-vars get sourced. Therefore accessing any of them results in "unbound variable" errors.

This is fixed by sourcing `$NIX_ATTRS_SH_FILE`, as-per 8bc5104a6e6a6acf09f9d7ddf1c6a9293efb405a
https://github.com/NixOS/nixpkgs/blob/ecd810c4cc90890229ef353a12fb384c47ee75c3/pkgs/stdenv/generic/default-builder.sh#L1

However `$__structuredAttrs` itself is still unbound, as it is not declared by that file. To work around this, I injected the value using `replaceVars`.

Once the main issue was resolved, I still had to re-work some implementation details due to differences in the `$outputs` variable in structured & non-structured attrs.

Without structuredAttrs, `$outputs` is an array of variable-names, which should be _dereferenced_ to access the actual path. However, with structuredAttrs, `$outputs` is an associative array:
```sh
declare -A outputs=(['out']='/nix/store/2z4mnrzxi1sb4k7zq49gibvvly7gs9rl-example' )
```

Tested using:
```nix
{
  structured = pkgs.testers.testBuildFailure (
    pkgs.runCommandLocal "example" { __structuredAttrs = true; } ''
      exit 1
    ''
  );
  unstructured = pkgs.testers.testBuildFailure (
    pkgs.runCommandLocal "example" { __structuredAttrs = false; } ''
      exit 1
    ''
  );
  unspecified = pkgs.testers.testBuildFailure (
    pkgs.runCommandLocal "example" { } ''
      exit 1
    ''
  );
}
```

</p>
</details>

Closes #355953

Also, uncomments a neovim test that was disabled in #352727 due to this bug

cc @wolfgangwalther @teto

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

MattSturgeon committed:
testers.testBuildFailure: fix structuredAttrs support

Fixes `testers.testBuildFailure` compatibility with `__structuredAttrs`.

The two main issues are, when `structuredAttrs` is enabled:
- `$__structuredAttrs` is not set
- `$outputs` is not set

`$__structuredAttrs` not being set is fixed by checking for
`$NIX_ATTRS_SH_FILE`, as per pkgs/stdenv/generic/setup.sh

`$outputs` not being set is fixed by sourcing `$NIX_ATTRS_SH_FILE`,
as-per pkgs/stdenv/generic/default-builder.sh

MattSturgeon committed:
neovim.tests: enable `failing_check`

teto reviewed (approved):
ty for the fix

wolfgangwalther reviewed (dismissed):
I will do a a real review later.

teto commented:
I forgot to mention it in my first message but your github ticket has a nice description of the issue, please copy it in the commit message. It helps when running `git blame` to know why the code is here.

MattSturgeon commented on pkgs/build-support/testers/default.nix:
```diff
@@ -24,7 +24,10 @@
+      (replaceVars ./expect-failure.sh {
+        coreutils = buildPackages.coreutils;
+        structuredAttrs = orig.__structuredAttrs or false;
+      })
```
I found the relevant impl in `setup.sh`, so I've managed to simplify this PR by adapting that: https://github.com/NixOS/nixpkgs/blob/a1ad5919e453255ea17fb2368113fd528ee4b29f/pkgs/stdenv/generic/setup.sh#L28-L57

teto commented:
I defer the review to the real boss @wolfgangwalther who made such huge contributions to structured attrs (thanks again). Happy with this PR goal.

MattSturgeon commented on pkgs/build-support/testers/expect-failure.sh:
```diff
@@ -35,7 +35,28 @@ echo "testBuildFailure: Original builder produced exit code: $r"
 #
-# # from stdenv setup.sh
+# # based on stdenv setup.sh
+if [ -f .attrs.sh ] || [[ -n "${NIX_ATTRS_SH_FILE:-}" ]] || [[ -n "${NIX_ATTRS_JSON_FILE:-}" ]]; then
```
```suggestion
if [ -f .attrs.sh ] || [[ -n "${NIX_ATTRS_SH_FILE:-}" ]]; then
```

I think by the time we get here, `NIX_ATTRS_JSON_FILE` has already been coerced into a `NIX_ATTRS_SH_FILE`?

---

In fact, I think this can actually be:
```suggestion
if [[ -n "${NIX_ATTRS_SH_FILE:-}" ]]; then
```
Since I believe that var is reliably defined in this context?

---

Not sure how best to go about testing this. Do you know of any derivations that'd run into edge-cases here?

MattSturgeon commented on pkgs/build-support/testers/expect-failure.sh:
```diff
@@ -35,7 +35,28 @@ echo "testBuildFailure: Original builder produced exit code: $r"
+    __structuredAttrs=1
+
+    # Source the attr-vars
+    if [ -e "${NIX_ATTRS_SH_FILE:-}" ]; then
+        . "$NIX_ATTRS_SH_FILE"
+    elif [ -f .attrs.sh ]; then
+        . .attrs.sh
+    fi
```
Is this reliably setup before running this script?

```suggestion
    if [ -e "$NIX_ATTRS_SH_FILE" ]; then
        . "$NIX_ATTRS_SH_FILE"
    fi
```

----

Or even?
```suggestion
    . "$NIX_ATTRS_SH_FILE"
```

MattSturgeon commented on pkgs/build-support/testers/expect-failure.sh:
```diff
@@ -35,7 +35,28 @@ echo "testBuildFailure: Original builder produced exit code: $r"
 # Write the build log to the default output
 #
-# # from stdenv setup.sh
+# # based on stdenv setup.sh
```
> Do not export any variables.

My current impl is definitely violating this. It is at least happening _after_ invoking the original builder.

MattSturgeon replied:
I guess we could just encapsulate sourcing the vars in a subshell though

MattSturgeon replied:
I've pushed this as a `fixup` commit. I can squash once @roberth has had chance to chime in.

roberth replied:
The goal is to keep the general environment as pristine as possible when entering the script, and avoid any side effects.
Unfortunately, `setup.sh` will run arbitrary code from the "wrapped" derivation - primarily `postHook` and `userHook`. These may well have side effects that affect more than the subshell variables.
Construed example:
```nix
postHook = ''
  [[ ! -e poison ]]
  touch poison
'';
```
Should've been a test case.

MattSturgeon replied:
> Unfortunately, `setup.sh` will run arbitrary code from the "wrapped" derivation - primarily `postHook` and `userHook`. These may well have side effects [...]

Gotcha, makes sense. I've removed sourceing the setup script. Instead, I'm passing in the ordered `outputs` list from the drv-args using `replaceVars` and `toShellVars`.

> Should've been a test case.

Added. I can verify the test fails when sourcing `setup`.

MattSturgeon commented:
Ideally, I'd add a regression test for this. But I don't see any existing tests; `tests.testers` is just an alias for `testers`...

Looking at the stdenv tests, that has a bunch of stuff I don't currently understand relating to `buildPackages` and getting an "early" instance of pkgs. Is any of that also required for testing `testers`?

Or is there a test file somewhere I missed?

MattSturgeon committed:
tests.testers.testBuildFailure: test with `structuredAttrs`

MattSturgeon commented:
> Ideally, I'd add a regression test for this. But I don't see any existing tests to add to; `tests.testers` is just an alias for `testers`...
> 
> Or is there a test file somewhere I missed?

Turns out I'm blind. `tests.testers` is not an alias, but is an actual test file.

I've added some basic tests that piggy-back on existing tests. However the multiOutput test is failing with structuredAttrs...

```
$ nix log /nix/store/b5x26d73b0nxxlh6anph4xj50f9lpnzi-testBuildFailure-multiOutput.drv
grep: /nix/store/3siwh00k6n76maya58scqizahw1pm31b-fail-structuredAttrs-dev/testBuildFailure.log: No such file or directory
```

I haven't had chance to debug yet, but my initial thoughts are:
Maybe the _default_ output is not guaranteed to be the _first_ output in the structured associative array?

MattSturgeon commented:
> associative arrays are hash-ordered and thus the order is not preserved.
> attrs.json should have the same information with the right order.

Yeah, I thought about that, but figured it'd be better to inject the head-output from the drv-args, rather than inject `jq` _and_ parse the json.

> It'd seem possibly useful to have stdenv set a variable to point at the default output, even for structuredAttrs.

Agreed. This would be the cleanest solution. It seems a bit beyond the scope of this PR, but maybe it'd be worth it...

roberth commented on pkgs/build-support/testers/expect-failure.sh:
```diff
@@ -34,18 +34,20 @@ echo "testBuildFailure: Original builder produced exit code: $r"
+# NOTE: This MUST be done after the original builder has finished!
+#       Otherwise we could pollute its environment.
+if [ -e "${NIX_ATTRS_SH_FILE:-}" ]; then . "$NIX_ATTRS_SH_FILE"; elif [ -f .attrs.sh ]; then . .attrs.sh; fi
+source "$stdenv/setup"
```
For now, could we copy the functionality we need?

Later, we could refactor `setup.sh` so that it is constructed at evaluation time with `builtins.toFile`. This also allows us to strip comments, so that we can improve those without mass rebuilds.

MattSturgeon committed:
tests.testers.testBuildFailure: assert hooks have no side-effects

If `$stdenv/setup` is sourced by `expect-failure.sh`, then the original
builder's hooks could be re-run, potentially leading to unintended
side-effects.

wolfgangwalther reviewed (approved):
Just a nit, very nice approach now!

MattSturgeon commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 371337`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 25 packages built:</summary>
  <ul>
    <li>tests.dotnet.structured-attrs.no-structured-attrs</li>
    <li>tests.dotnet.use-dotnet-from-env.without-fallback</li>
    <li>tests.replaceVars.replaceVars-fails-in-build-phase</li>
    <li>tests.replaceVars.replaceVars-fails-in-check-phase</li>
    <li>tests.replaceVars.replaceVars-fails-in-check-phase-with-exemption</li>
    <li>tests.replaceVars.replaceVars-fails-on-directory</li>
    <li>tests.testers.hasPkgConfigModules.miniz-versions-mismatch</li>
    <li>tests.testers.hasPkgConfigModules.zlib-does-not-have-ylib</li>
    <li>tests.testers.lycheeLinkCheck.fail</li>
    <li>tests.testers.lycheeLinkCheck.fail-emptyDirectory</li>
    <li>tests.testers.shellcheck.example-dir</li>
    <li>tests.testers.shellcheck.example-file</li>
    <li>tests.testers.testBuildFailure.happy</li>
    <li>tests.testers.testBuildFailure.happyStructuredAttrs</li>
    <li>tests.testers.testBuildFailure.helloDoesNotFail</li>
    <li>tests.testers.testBuildFailure.multiOutput</li>
    <li>tests.testers.testBuildFailure.multiOutputStructuredAttrs</li>
    <li>tests.testers.testBuildFailure.sideEffectStructuredAttrs</li>
    <li>tests.testers.testBuildFailure.sideEffects</li>
    <li>tests.testers.testEqualContents.emptyFileAndDir</li>
    <li>tests.testers.testEqualContents.fileDiff</li>
    <li>tests.testers.testEqualContents.fileMissing</li>
    <li>tests.testers.testEqualContents.nonExistentPath</li>
    <li>tests.testers.testEqualContents.unequalExe</li>
    <li>tests.testers.testEqualContents.unequalExeInDir</li>
  </ul>
</details>
